File: 35-search-insert-position/search-insert-position.java
class Solution {
    public int binSearch(int[] arr, int left, int right, int target) {
        if (left > right) {
            return left; // Return the position to insert if not found
        }
        int mid = left + (right - left) / 2; // Calculate mid to avoid overflow

        if (arr[mid] == target) {
            return mid; // Target found
        }
        if (arr[mid] > target) {
            return binSearch(arr, left, mid - 1, target); // Search in the left half
        } else {
            return binSearch(arr, mid + 1, right, target); // Search in the right half
        }
    }

    public int searchInsert(int[] nums, int target) {
        return binSearch(nums, 0, nums.length - 1, target);
    }
}

File: 35-search-insert-position/search-insert-position.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        # Finding the ceiling 

        start = 0
        end = len(nums) - 1
        res = -1
        while(start <= end):
            mid = (start+end)//2
            if target == nums[mid]:
                return mid
            elif target > nums[mid]:
                res = mid
                start = mid+1
            else:
                end = mid-1
        return res+1
